File: third-party/thrift/src/thrift/lib/go/thrift/debug_protocol.go
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package thrift

import (
	"log"
)

type DebugProtocol struct {
	Delegate  Protocol
	LogPrefix string
}

type DebugProtocolFactory struct {
	Underlying ProtocolFactory
	LogPrefix  string
}

func NewDebugProtocolFactory(underlying ProtocolFactory, logPrefix string) *DebugProtocolFactory {
	return &DebugProtocolFactory{
		Underlying: underlying,
		LogPrefix:  logPrefix,
	}
}

func (t *DebugProtocolFactory) GetProtocol(trans Transport) Protocol {
	return &DebugProtocol{
		Delegate:  t.Underlying.GetProtocol(trans),
		LogPrefix: t.LogPrefix,
	}
}

func (tdp *DebugProtocol) WriteMessageBegin(name string, typeId MessageType, seqid int32) error {
	err := tdp.Delegate.WriteMessageBegin(name, typeId, seqid)
	log.Printf("%sWriteMessageBegin(name=%#v, typeId=%#v, seqid=%#v) => %#v", tdp.LogPrefix, name, typeId, seqid, err)
	return err
}
func (tdp *DebugProtocol) WriteMessageEnd() error {
	err := tdp.Delegate.WriteMessageEnd()
	log.Printf("%sWriteMessageEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteStructBegin(name string) error {
	err := tdp.Delegate.WriteStructBegin(name)
	log.Printf("%sWriteStructBegin(name=%#v) => %#v", tdp.LogPrefix, name, err)
	return err
}
func (tdp *DebugProtocol) WriteStructEnd() error {
	err := tdp.Delegate.WriteStructEnd()
	log.Printf("%sWriteStructEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteFieldBegin(name string, typeId Type, id int16) error {
	err := tdp.Delegate.WriteFieldBegin(name, typeId, id)
	log.Printf("%sWriteFieldBegin(name=%#v, typeId=%#v, id%#v) => %#v", tdp.LogPrefix, name, typeId, id, err)
	return err
}
func (tdp *DebugProtocol) WriteFieldEnd() error {
	err := tdp.Delegate.WriteFieldEnd()
	log.Printf("%sWriteFieldEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteFieldStop() error {
	err := tdp.Delegate.WriteFieldStop()
	log.Printf("%sWriteFieldStop() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteMapBegin(keyType Type, valueType Type, size int) error {
	err := tdp.Delegate.WriteMapBegin(keyType, valueType, size)
	log.Printf("%sWriteMapBegin(keyType=%#v, valueType=%#v, size=%#v) => %#v", tdp.LogPrefix, keyType, valueType, size, err)
	return err
}
func (tdp *DebugProtocol) WriteMapEnd() error {
	err := tdp.Delegate.WriteMapEnd()
	log.Printf("%sWriteMapEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteListBegin(elemType Type, size int) error {
	err := tdp.Delegate.WriteListBegin(elemType, size)
	log.Printf("%sWriteListBegin(elemType=%#v, size=%#v) => %#v", tdp.LogPrefix, elemType, size, err)
	return err
}
func (tdp *DebugProtocol) WriteListEnd() error {
	err := tdp.Delegate.WriteListEnd()
	log.Printf("%sWriteListEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteSetBegin(elemType Type, size int) error {
	err := tdp.Delegate.WriteSetBegin(elemType, size)
	log.Printf("%sWriteSetBegin(elemType=%#v, size=%#v) => %#v", tdp.LogPrefix, elemType, size, err)
	return err
}
func (tdp *DebugProtocol) WriteSetEnd() error {
	err := tdp.Delegate.WriteSetEnd()
	log.Printf("%sWriteSetEnd() => %#v", tdp.LogPrefix, err)
	return err
}
func (tdp *DebugProtocol) WriteBool(value bool) error {
	err := tdp.Delegate.WriteBool(value)
	log.Printf("%sWriteBool(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteByte(value byte) error {
	err := tdp.Delegate.WriteByte(value)
	log.Printf("%sWriteByte(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteI16(value int16) error {
	err := tdp.Delegate.WriteI16(value)
	log.Printf("%sWriteI16(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteI32(value int32) error {
	err := tdp.Delegate.WriteI32(value)
	log.Printf("%sWriteI32(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteI64(value int64) error {
	err := tdp.Delegate.WriteI64(value)
	log.Printf("%sWriteI64(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteDouble(value float64) error {
	err := tdp.Delegate.WriteDouble(value)
	log.Printf("%sWriteDouble(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteFloat(value float32) error {
	err := tdp.Delegate.WriteFloat(value)
	log.Printf("%sWriteFloat(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteString(value string) error {
	err := tdp.Delegate.WriteString(value)
	log.Printf("%sWriteString(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}
func (tdp *DebugProtocol) WriteBinary(value []byte) error {
	err := tdp.Delegate.WriteBinary(value)
	log.Printf("%sWriteBinary(value=%#v) => %#v", tdp.LogPrefix, value, err)
	return err
}

func (tdp *DebugProtocol) ReadMessageBegin() (name string, typeId MessageType, seqid int32, err error) {
	name, typeId, seqid, err = tdp.Delegate.ReadMessageBegin()
	log.Printf("%sReadMessageBegin() (name=%#v, typeId=%#v, seqid=%#v, err=%#v)", tdp.LogPrefix, name, typeId, seqid, err)
	return
}
func (tdp *DebugProtocol) ReadMessageEnd() (err error) {
	err = tdp.Delegate.ReadMessageEnd()
	log.Printf("%sReadMessageEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadStructBegin() (name string, err error) {
	name, err = tdp.Delegate.ReadStructBegin()
	log.Printf("%sReadStructBegin() (name%#v, err=%#v)", tdp.LogPrefix, name, err)
	return
}
func (tdp *DebugProtocol) ReadStructEnd() (err error) {
	err = tdp.Delegate.ReadStructEnd()
	log.Printf("%sReadStructEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadFieldBegin() (name string, typeId Type, id int16, err error) {
	name, typeId, id, err = tdp.Delegate.ReadFieldBegin()
	log.Printf("%sReadFieldBegin() (name=%#v, typeId=%#v, id=%#v, err=%#v)", tdp.LogPrefix, name, typeId, id, err)
	return
}
func (tdp *DebugProtocol) ReadFieldEnd() (err error) {
	err = tdp.Delegate.ReadFieldEnd()
	log.Printf("%sReadFieldEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadMapBegin() (keyType Type, valueType Type, size int, err error) {
	keyType, valueType, size, err = tdp.Delegate.ReadMapBegin()
	log.Printf("%sReadMapBegin() (keyType=%#v, valueType=%#v, size=%#v, err=%#v)", tdp.LogPrefix, keyType, valueType, size, err)
	return
}
func (tdp *DebugProtocol) ReadMapEnd() (err error) {
	err = tdp.Delegate.ReadMapEnd()
	log.Printf("%sReadMapEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadListBegin() (elemType Type, size int, err error) {
	elemType, size, err = tdp.Delegate.ReadListBegin()
	log.Printf("%sReadListBegin() (elemType=%#v, size=%#v, err=%#v)", tdp.LogPrefix, elemType, size, err)
	return
}
func (tdp *DebugProtocol) ReadListEnd() (err error) {
	err = tdp.Delegate.ReadListEnd()
	log.Printf("%sReadListEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadSetBegin() (elemType Type, size int, err error) {
	elemType, size, err = tdp.Delegate.ReadSetBegin()
	log.Printf("%sReadSetBegin() (elemType=%#v, size=%#v, err=%#v)", tdp.LogPrefix, elemType, size, err)
	return
}
func (tdp *DebugProtocol) ReadSetEnd() (err error) {
	err = tdp.Delegate.ReadSetEnd()
	log.Printf("%sReadSetEnd() err=%#v", tdp.LogPrefix, err)
	return
}
func (tdp *DebugProtocol) ReadBool() (value bool, err error) {
	value, err = tdp.Delegate.ReadBool()
	log.Printf("%sReadBool() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadByte() (value byte, err error) {
	value, err = tdp.Delegate.ReadByte()
	log.Printf("%sReadByte() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadI16() (value int16, err error) {
	value, err = tdp.Delegate.ReadI16()
	log.Printf("%sReadI16() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadI32() (value int32, err error) {
	value, err = tdp.Delegate.ReadI32()
	log.Printf("%sReadI32() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadI64() (value int64, err error) {
	value, err = tdp.Delegate.ReadI64()
	log.Printf("%sReadI64() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadDouble() (value float64, err error) {
	value, err = tdp.Delegate.ReadDouble()
	log.Printf("%sReadDouble() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadFloat() (value float32, err error) {
	value, err = tdp.Delegate.ReadFloat()
	log.Printf("%sReadFloat() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadString() (value string, err error) {
	value, err = tdp.Delegate.ReadString()
	log.Printf("%sReadString() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) ReadBinary() (value []byte, err error) {
	value, err = tdp.Delegate.ReadBinary()
	log.Printf("%sReadBinary() (value=%#v, err=%#v)", tdp.LogPrefix, value, err)
	return
}
func (tdp *DebugProtocol) Skip(fieldType Type) (err error) {
	err = tdp.Delegate.Skip(fieldType)
	log.Printf("%sSkip(fieldType=%#v) (err=%#v)", tdp.LogPrefix, fieldType, err)
	return
}
func (tdp *DebugProtocol) Flush() (err error) {
	err = tdp.Delegate.Flush()
	log.Printf("%sFlush() (err=%#v)", tdp.LogPrefix, err)
	return
}

// Deprecated: Transport() is a deprecated method.
func (tdp *DebugProtocol) Transport() Transport {
	return tdp.Delegate.Transport()
}

func (tdp *DebugProtocol) Close() error {
	return tdp.Delegate.Close()
}

File: third-party/thrift/src/thrift/lib/cpp2/protocol/DebugProtocol.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <thrift/lib/cpp2/protocol/DebugProtocol.h>

#include <folly/Conv.h>
#include <folly/String.h>

FOLLY_GFLAGS_DEFINE_bool(
    thrift_cpp2_debug_skip_list_indices,
    false,
    "Wether to skip indices when debug-printing lists (unless overridden)");
FOLLY_GFLAGS_DEFINE_int64(
    thrift_cpp2_debug_string_limit,
    256,
    "Limit on string size when debug-printing thrift, 0 is no limit");

namespace apache::thrift {

DebugProtocolWriter::DebugProtocolWriter(
    ExternalBufferSharing /*sharing*/, Options options)
    : out_(nullptr, 0), options_(options) {}

namespace {

std::string fieldTypeName(TType type) {
  switch (type) {
    case TType::T_STOP:
      return "stop";
    case TType::T_VOID:
      return "void";
    case TType::T_BOOL:
      return "bool";
    case TType::T_BYTE:
      return "byte";
    case TType::T_I16:
      return "i16";
    case TType::T_I32:
      return "i32";
    case TType::T_U64:
      return "u64";
    case TType::T_I64:
      return "i64";
    case TType::T_DOUBLE:
      return "double";
    case TType::T_FLOAT:
      return "float";
    case TType::T_STRING:
      return "string";
    case TType::T_STRUCT:
      return "struct";
    case TType::T_MAP:
      return "map";
    case TType::T_SET:
      return "set";
    case TType::T_LIST:
      return "list";
    case TType::T_UTF8:
      return "utf8";
    case TType::T_UTF16:
      return "utf16";
    case TType::T_STREAM:
      return "stream";
    default:
      return fmt::format("unknown({})", int(type));
  }
}

const size_t kIndent = 2;

} // namespace

void DebugProtocolWriter::indentUp() {
  indent_.append(kIndent, ' ');
}

void DebugProtocolWriter::indentDown() {
  CHECK_GE(indent_.size(), kIndent);
  indent_.erase(indent_.size() - kIndent);
}

void DebugProtocolWriter::pushState(ItemType t) {
  indentUp();
  writeState_.emplace_back(t);
}

void DebugProtocolWriter::popState() {
  CHECK(!writeState_.empty());
  writeState_.pop_back();
  indentDown();
}

void DebugProtocolWriter::startItem() {
  if (writeState_.empty()) { // top level
    return;
  }
  auto& ws = writeState_.back();
  switch (ws.type) {
    case STRUCT:
      break;
    case SET:
    case MAP_KEY:
      writeIndent();
      break;
    case MAP_VALUE:
      writePlain(" -> ");
      break;
    case LIST:
      if (options_.skipListIndices) {
        writeIndent();
      } else {
        writeIndented("[{}] = ", ws.index);
      }
      break;
  }
}

void DebugProtocolWriter::endItem() {
  if (writeState_.empty()) { // top level
    return;
  }
  auto& ws = writeState_.back();
  ++ws.index;
  switch (ws.type) {
    case LIST:
    case STRUCT:
    case SET:
      writePlain(",\n");
      break;
    case MAP_KEY:
      ws.type = MAP_VALUE;
      break;
    case MAP_VALUE:
      ws.type = MAP_KEY;
      writePlain(",\n");
  }
}

void DebugProtocolWriter::setOutput(
    folly::IOBufQueue* storage, size_t maxGrowth) {
  // Allocate 16KB at a time; leave some room for the IOBuf overhead
  constexpr size_t kDesiredGrowth = (1 << 14) - 64;
  out_.reset(storage, std::min(kDesiredGrowth, maxGrowth));
}

void DebugProtocolWriter::setOutput(folly::io::QueueAppender&& output) {
  out_ = std::move(output);
}

uint32_t DebugProtocolWriter::writeMessageBegin(
    const std::string& name, MessageType messageType, int32_t /*seqid*/) {
  std::string mtype;
  switch (messageType) {
    case MessageType::T_CALL:
      mtype = "call";
      break;
    case MessageType::T_REPLY:
      mtype = "reply";
      break;
    case MessageType::T_EXCEPTION:
      mtype = "exn";
      break;
    case MessageType::T_ONEWAY:
      mtype = "oneway";
      break;
  }

  writeIndented("({}) {}(", mtype, name);
  indentUp();
  return 0;
}

uint32_t DebugProtocolWriter::writeMessageEnd() {
  indentDown();
  writeIndented(")\n");
  return 0;
}

uint32_t DebugProtocolWriter::writeStructBegin(const char* name) {
  startItem();
  writePlain("{} {{\n", name);
  pushState(STRUCT);
  return 0;
}

uint32_t DebugProtocolWriter::writeStructEnd() {
  popState();
  writeIndented("}}");
  endItem();
  return 0;
}

uint32_t DebugProtocolWriter::writeFieldBegin(
    const char* name, TType fieldType, int16_t fieldId) {
  writeIndent();
  if (!options_.skipFieldId) {
    writePlain("{:0d}: ", fieldId);
  }
  writePlain("{}", name);
  if (!options_.skipFieldType) {
    writePlain(" ({})", fieldTypeName(fieldType));
  }
  writePlain(" = ");
  return 0;
}

uint32_t DebugProtocolWriter::writeFieldEnd() {
  return 0;
}
uint32_t DebugProtocolWriter::writeFieldStop() {
  return 0;
}

uint32_t DebugProtocolWriter::writeMapBegin(
    TType keyType, TType valueType, uint32_t size) {
  startItem();
  writePlain(
      "map<{},{}>[{}] {{\n",
      fieldTypeName(keyType),
      fieldTypeName(valueType),
      size);
  pushState(MAP_KEY);
  return 0;
}

uint32_t DebugProtocolWriter::writeMapEnd() {
  popState();
  writeIndented("}}");
  endItem();
  return 0;
}

uint32_t DebugProtocolWriter::writeListBegin(TType elemType, uint32_t size) {
  startItem();
  writePlain("list<{}>[{}] {{\n", fieldTypeName(elemType), size);
  pushState(LIST);
  return 0;
}

uint32_t DebugProtocolWriter::writeListEnd() {
  popState();
  writeIndented("}}");
  endItem();
  return 0;
}

uint32_t DebugProtocolWriter::writeSetBegin(TType elemType, uint32_t size) {
  startItem();
  writePlain("set<{}>[{}] {{\n", fieldTypeName(elemType), size);
  pushState(SET);
  return 0;
}

uint32_t DebugProtocolWriter::writeSetEnd() {
  popState();
  writeIndented("}}");
  endItem();
  return 0;
}

uint32_t DebugProtocolWriter::writeBool(bool v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeByte(int8_t v) {
  writeItem("0x{:x}", uint8_t(v));
  return 0;
}

uint32_t DebugProtocolWriter::writeI16(int16_t v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeI32(int32_t v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeI64(int64_t v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeFloat(float v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeDouble(double v) {
  writeItem("{}", v);
  return 0;
}

uint32_t DebugProtocolWriter::writeString(folly::StringPiece str) {
  return writeBinary(str);
}

uint32_t DebugProtocolWriter::writeBinary(folly::StringPiece str) {
  return writeBinary(folly::ByteRange(str));
}

uint32_t DebugProtocolWriter::writeBinary(folly::ByteRange v) {
  writeByteRange(v);
  return 0;
}

uint32_t DebugProtocolWriter::writeBinary(
    const std::unique_ptr<folly::IOBuf>& str) {
  if (str) {
    writeByteRange(folly::ByteRange(str->clone()->coalesce()));
  }
  return 0;
}

uint32_t DebugProtocolWriter::writeBinary(const folly::IOBuf& str) {
  writeByteRange(folly::ByteRange(str.clone()->coalesce()));
  return 0;
}

uint32_t DebugProtocolWriter::serializedMessageSize(
    const std::string& /*name*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedFieldSize(
    const char* /*name*/, TType /*fieldName*/, int16_t /*fieldId*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedStructSize(const char* /*name*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeMapBegin(
    TType /*keyType*/, TType /*valType*/, uint32_t /*size*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeMapEnd() {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeListBegin(
    TType /*elemType*/, uint32_t /*size*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeListEnd() {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeSetBegin(
    TType /*elemType*/, uint32_t /*size*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeSetEnd() {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeStop() {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeBool(bool /*value*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeByte(int8_t /*byte*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeI16(int16_t /*i16*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeI32(int32_t /*i32*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeI64(int64_t /*i64*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeDouble(double /*dub*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeFloat(float /*flt*/) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeString(folly::StringPiece str) {
  return serializedSizeBinary(str);
}

uint32_t DebugProtocolWriter::serializedSizeBinary(folly::StringPiece str) {
  return serializedSizeBinary(folly::ByteRange(str));
}

uint32_t DebugProtocolWriter::serializedSizeBinary(folly::ByteRange) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeBinary(
    const std::unique_ptr<folly::IOBuf>&) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeBinary(const folly::IOBuf&) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeZCBinary(folly::StringPiece str) {
  return serializedSizeBinary(folly::ByteRange(str));
}

uint32_t DebugProtocolWriter::serializedSizeZCBinary(folly::ByteRange) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeZCBinary(
    const std::unique_ptr<folly::IOBuf>&) {
  return 0;
}

uint32_t DebugProtocolWriter::serializedSizeZCBinary(const folly::IOBuf&) {
  return 0;
}

void DebugProtocolWriter::writeByteRange(folly::ByteRange v) {
  static constexpr size_t kStringPrefixSize = 128;
  const size_t limit = options_.stringLengthLimit;

  auto str = folly::StringPiece(v);

  folly::StringPiece toShow = limit != 0 && str.size() > limit
      ? str.subpiece(0, std::min(kStringPrefixSize, limit))
      : str;
  std::string printed = folly::cEscape<std::string>(toShow);
  if (toShow.size() < str.size()) {
    folly::toAppend("[...](", str.size(), ")", &printed);
  }

  writeItem("\"{}\"", printed);
}

} // namespace apache::thrift
